### PR TITLE
example configurations for UI choices

### PR DESCRIPTION
**Pull request by minrk**
show how to use classic notebook, new server, juptyerlab, and retrolab.

- consistent config, works for both 1.x and 2.x, no matter what the default is
- examples covered: notebook, lab, lab on legacy server, nbclassic, and retrolab

cf #776

### CODE REVIEW

**Review by consideRatio — APPROVED** [review]
Wieee excellent job on this! It looks good to me! I'll leave it open for a day at least for additional feedback.

File: doc/source/jupyterhub/customizing/user-environment.md [hunk 1]
```diff
@@ -69,28 +69,157 @@ If you'd like users to select an environment from **multiple docker images**,
...
+(user-interfaces)=
+
+## Selecting a user interface
+
+[JupyterLab][] is the new user interface for Jupyter,
+which is meant to replace the classic notebook user interface (UI).
+While users already can interchange `/tree` and `/lab` in the URL to switch between
+the classic UI and JupyterLab if both are installed,
+Deployments using JupyterHub 1.x and earlier default to the classic UI,
+while 2.0 will make JupyterLab the default.
```
Comment (manics) [inline doc/source/jupyterhub/customizing/user-environment.md:81]:
```suggestion
Users already can interchange `/tree` and `/lab` in the URL to switch between
the classic UI and JupyterLab if both are installed.
Deployments using JupyterHub 1.x and earlier default to the classic UI,
while JupyterHub 2.0 makes JupyterLab the default.
```

File: doc/source/jupyterhub/customizing/user-environment.md [hunk 2]
```diff
@@ -69,28 +69,157 @@ If you'd like users to select an environment from **multiple docker images**,
...
 (jupyterlab-by-default)=
 
 ## Use JupyterLab by default
 
-[JupyterLab](https://jupyterlab.readthedocs.io/en/stable/index.html) is a new
-user interface for Jupyter about to replace the classic user interface (UI).
-While users already can interchange `/tree` and `/lab` in the URL to switch between
-the classic UI and JupyterLab, they will default to use the classic UI.
+:::{note}
+This will be the default in JupyterHub 2.0 and helm chart 2.0.
```
Comment (manics) [inline doc/source/jupyterhub/customizing/user-environment.md:142]:
```suggestion
This is the default in JupyterHub 2.0 and Helm chart 2.0.
```

File: doc/source/jupyterhub/customizing/user-environment.md [hunk 3]
```diff
@@ -69,28 +69,157 @@ If you'd like users to select an environment from **multiple docker images**,
...
+    JUPYTERHUB_SINGLEUSER_APP: "jupyter_server.serverapp.ServerApp"
+```
+
+### Alternative interfaces
+
+There are more Jupyter server extensions providing alternate UI choices,
+which can be used with JupyterHub.
+
+For example, [retrolab][] is a different, built on JupyterLab,
+but which may be more comfortable for those coming from the classic Jupyter UI.
```
Comment (manics) [inline doc/source/jupyterhub/customizing/user-environment.md:207]:
```suggestion
For example, [retrolab][] is a different notebook interface, built on JupyterLab,
but which may be more comfortable for those coming from the classic Jupyter UI.
```

File: doc/source/jupyterhub/customizing/user-environment.md [hunk 4]
```diff
@@ -69,28 +69,157 @@ If you'd like users to select an environment from **multiple docker images**,
...
+
+[retrolab]: https://blog.jupyter.org/retrolab-a-jupyterlab-distribution-with-a-retro-look-and-feel-8096b8b223d0
+
+To install such an extension:
+
+1. install the package (`pip install retrolab` or `conda install retrolab`) in your user image
```
Comment (manics) [inline doc/source/jupyterhub/customizing/user-environment.md:213]:
```suggestion
1. install the package (`pip install retrolab` or `conda install retrolab`) in your user container image
```

File: doc/source/jupyterhub/customizing/user-environment.md [hunk 5]
```diff
@@ -69,28 +69,157 @@ If you'd like users to select an environment from **multiple docker images**,
...
+[jupyterlab]: https://jupyterlab.readthedocs.io
+
+There are two things to customize, when picking the user interface to launch for users:
+
+1. the user interface (UI)
+2. the server program to launch
+
+There are two main Jupyter server implementations.
+_Most_ deployments will not see a difference,
+but there can be issues for certain server extensions.
```
Comment (manics) [inline doc/source/jupyterhub/customizing/user-environment.md:92]: Do you know of a good external page we could link to that explains the potential problems?

**Review by manics — COMMENTED** [review]
A few small comments but sounds fine to me.

File: doc/source/jupyterhub/customizing/user-environment.md (hunk 1 above)
Comment (consideRatio) [reply doc/source/jupyterhub/customizing/user-environment.md:81]: Are both these variations reasonable English? The latter feels more right to me.

`Users already can interchange` or `Users can already interchange`

Comment (manics) [reply doc/source/jupyterhub/customizing/user-environment.md:81]: The latter sounds better, I'll update my suggestion.

File: doc/source/jupyterhub/customizing/user-environment.md (hunk 5 above)
Comment (minrk) [reply doc/source/jupyterhub/customizing/user-environment.md:92]: I don't know of good docs covering what is likely to break, other than possible use of private APIs or assumptions not preserved across the transition. Most things are meant to work, and [this doc][] describes migrating an extension. I don't actually know of any examples of classic server extensions that don't work today with the new server. Maybe I should be softer and say something more like:

> Jupyter Server aims to be as backward-compatible as possible,
> so the vast majority of server extensions continue to work across the upgrade.
> If you experience issues, however, please report them to the Jupyter Server team,
> and you can use the classic server in the meantime.


[this doc]: https://jupyter-server.readthedocs.io/en/latest/developers/extensions.html#migrating-an-extension-to-use-jupyter-server

File: doc/source/jupyterhub/customizing/user-environment.md [hunk 6]
```diff
@@ -69,28 +69,157 @@ If you'd like users to select an environment from **multiple docker images**,
...
+[JupyterLab][] is the new user interface for Jupyter,
+which is meant to replace the classic notebook user interface (UI).
+Users can already interchange `/tree` and `/lab` in the URL to switch between
+the classic UI and JupyterLab if both are installed.
+Deployments using JupyterHub 1.x and earlier default to the classic UI,
+while JupyterHub 2.0 makes JupyterLab the default.
+
+[jupyterlab]: https://jupyterlab.readthedocs.io
+
+There are two things to customize, when picking the user interface to launch for users:
```
Comment (willingc) [inline doc/source/jupyterhub/customizing/user-environment.md:85]:
```suggestion
To pick a user interface to launch by default for users, two customization items need to be set:
```

File: doc/source/jupyterhub/customizing/user-environment.md [hunk 7]
```diff
@@ -69,28 +69,157 @@ If you'd like users to select an environment from **multiple docker images**,
...
+
+1. the user interface (UI)
```
Comment (willingc) [inline doc/source/jupyterhub/customizing/user-environment.md:87]:
```suggestion
1. the preferred default user interface (UI)
```

File: doc/source/jupyterhub/customizing/user-environment.md [hunk 8]
```diff
@@ -69,28 +69,157 @@ If you'd like users to select an environment from **multiple docker images**,
...
+2. the server program to launch
+
+There are two main Jupyter server implementations.
```
Comment (willingc) [inline doc/source/jupyterhub/customizing/user-environment.md:90]:
```suggestion
There are two main Jupyter server implementations
```

File: doc/source/jupyterhub/customizing/user-environment.md [hunk 9]
```diff
@@ -69,28 +69,157 @@ If you'd like users to select an environment from **multiple docker images**,
...
+_Most_ deployments will not see a difference,
```
Comment (willingc) [inline doc/source/jupyterhub/customizing/user-environment.md:91]:
```suggestion
(_Most deployments will not see a difference,
```

File: doc/source/jupyterhub/customizing/user-environment.md (hunk 5 above)
Comment (willingc) [reply doc/source/jupyterhub/customizing/user-environment.md:92]:
```suggestion
but there can be issues for certain server extensions. If unsure, new applications should choose `jupyter_server`._):
```

File: doc/source/jupyterhub/customizing/user-environment.md [hunk 10]
```diff
@@ -69,28 +69,157 @@ If you'd like users to select an environment from **multiple docker images**,
...
+
+and the default server by:
+
+```yaml
+singleuser:
+  extraEnv:
+    JUPYTERHUB_SINGLEUSER_APP: "..."
+```
+
+To select the modern server:
```
Comment (willingc) [inline doc/source/jupyterhub/customizing/user-environment.md:113]:
```suggestion
Specifically, use one of these options to select the modern server:
```

**Review by willingc — APPROVED** [review]
Thanks @minrk.

**Comment by consideRatio** [comment]
Thank you @minrk @manics @willingc!!! ❤️ 🎉 🌻

**Pull request merged**